File: java/Samples/Biometrics/CS/LatentFingerprintSampleCS/ResolutionForm.cs
using System;
using System.Drawing;
using System.Windows.Forms;

namespace Neurotec.Samples
{
	public partial class ResolutionForm : Form
	{
		#region Private Variables

		private Point _startPt;
		private Point _endPt;
		private bool _dragging;

		private const int EndMarkerSize = 4;

		#endregion Private Variables

		#region Constructor
		public ResolutionForm()
		{
			InitializeComponent();
		}
		#endregion

		#region Public Properties
		public float HorzResolution
		{
			get
			{
				return Convert.ToSingle(nudHorzResolution.Value);
			}
			set
			{
				decimal val = Convert.ToDecimal(value);
				if (val < nudHorzResolution.Minimum)
				{
					val = nudHorzResolution.Minimum;
				}
				else if (val > nudHorzResolution.Maximum)
				{
					val = nudHorzResolution.Maximum;
				}
				nudHorzResolution.Value = val;
			}
		}

		public float VertResolution
		{
			get
			{
				return Convert.ToSingle(nudVertResolution.Value);
			}
			set
			{
				decimal val = Convert.ToDecimal(value);
				if (val < nudVertResolution.Minimum)
				{
					val = nudVertResolution.Minimum;
				}
				else if (val > nudVertResolution.Maximum)
				{
					val = nudVertResolution.Maximum;
				}
				nudVertResolution.Value = val;
			}
		}

		public Bitmap FingerImage
		{
			set
			{
				pbFingerprint.Image = value;
			}
		}

		#endregion Public Properties

		#region Event Handling

		private float CalculateDistance()
		{
			return Convert.ToSingle(Math.Sqrt((_startPt.X - _endPt.X) * (double)(_startPt.X - _endPt.X) + (_startPt.Y - _endPt.Y) * (double)(_startPt.Y - _endPt.Y)));
		}

		private void pbFingerprint_MouseDown(object sender, MouseEventArgs e)
		{
			if (e.Button == MouseButtons.Left)
			{
				_startPt = e.Location;
				_endPt = Point.Empty;
				pbFingerprint.Cursor = Cursors.Cross;
				_dragging = true;
			}
		}

		private void pbFingerprint_MouseMove(object sender, MouseEventArgs e)
		{
			if (e.Button == MouseButtons.Left
				&& _dragging)
			{
				_endPt = e.Location;
				pbFingerprint.Invalidate();
			}
		}

		private void pbFingerprint_MouseUp(object sender, MouseEventArgs e)
		{
			if (e.Button == MouseButtons.Left
				&& _dragging)
			{
				_endPt = e.Location;
				Cursor = Cursors.Arrow;
				_dragging = false;

				float scaleToInch = 1.0f;
				if (rbCentimeters.Checked)
				{
					scaleToInch = 1.0f / 2.54f;
				}

				float distance = CalculateDistance();
				if (distance > 50.0)
				{
					nudHorzResolution.Value = Convert.ToDecimal(distance / (Convert.ToSingle(nudUnitScale.Value) * scaleToInch));
					nudVertResolution.Value = nudHorzResolution.Value;
				}
				else
				{
					Utilities.ShowInformation(@"Please draw a longer line segment.");
				}

				_startPt = Point.Empty;
				_endPt = Point.Empty;
				pbFingerprint.Invalidate();
			}
		}

		private void DrawSelectionLine(Graphics g, Pen pen)
		{
			// draw pointer markers
			g.DrawLine(pen, _startPt.X - EndMarkerSize, _startPt.Y, _startPt.X + EndMarkerSize, _startPt.Y);
			g.DrawLine(pen, _startPt.X, _startPt.Y - EndMarkerSize, _startPt.X, _startPt.Y + EndMarkerSize);
			g.DrawLine(pen, _endPt.X - EndMarkerSize, _endPt.Y, _endPt.X + EndMarkerSize, _endPt.Y);
			g.DrawLine(pen, _endPt.X, _endPt.Y - EndMarkerSize, _endPt.X, _endPt.Y + EndMarkerSize);
			// draw line between starting and ending points
			g.DrawLine(pen, _startPt, _endPt);
		}

		private void pbFingerprint_Paint(object sender, PaintEventArgs e)
		{
			if (!_startPt.IsEmpty
				&& !_endPt.IsEmpty)
			{
				Graphics g = e.Graphics;

				System.Drawing.Drawing2D.SmoothingMode smoothingMode = g.SmoothingMode;
				g.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality;

				using (var pen = new Pen(Color.White, 3))
				{
					DrawSelectionLine(g, pen);
				}

				using (var pen = new Pen(Color.Green, 1))
				{
					DrawSelectionLine(g, pen);
				}

				g.SmoothingMode = smoothingMode;
			}
		}

		private void nudHorzResolution_ValueChanged(object sender, EventArgs e)
		{
			errorProvider1.SetError(nudHorzResolution,
				nudHorzResolution.Value < 250m ? "Current resolution is lower than recommended minimum." : "");
		}

		private void nudVertResolution_ValueChanged(object sender, EventArgs e)
		{
			errorProvider1.SetError(nudVertResolution,
				nudVertResolution.Value < 250m ? "Current resolution is lower than recommended minimum." : "");
		}

		#endregion Event Handling
	}
}

File: java/Samples/Biometrics/VB/LatentFingerprintSampleVB/ResolutionForm.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Partial Public Class ResolutionForm
	Inherits Form
	#Region "Private Variables"

	Private _startPt As Point
	Private _endPt As Point
	Private _dragging As Boolean

	Private Const EndMarkerSize As Integer = 4

	#End Region ' Private Variables

	#Region "Constructor"
	Public Sub New()
		InitializeComponent()
	End Sub
	#End Region

	#Region "Public Properties"
	Public Property HorzResolution() As Single
		Get
			Return Convert.ToSingle(nudHorzResolution.Value)
		End Get
		Set(ByVal value As Single)
			Dim val As Decimal = Convert.ToDecimal(value)
			If val < nudHorzResolution.Minimum Then
				val = nudHorzResolution.Minimum
			ElseIf val > nudHorzResolution.Maximum Then
				val = nudHorzResolution.Maximum
			End If
			nudHorzResolution.Value = val
		End Set
	End Property

	Public Property VertResolution() As Single
		Get
			Return Convert.ToSingle(nudVertResolution.Value)
		End Get
		Set(ByVal value As Single)
			Dim val As Decimal = Convert.ToDecimal(value)
			If val < nudVertResolution.Minimum Then
				val = nudVertResolution.Minimum
			ElseIf val > nudVertResolution.Maximum Then
				val = nudVertResolution.Maximum
			End If
			nudVertResolution.Value = val
		End Set
	End Property

	Public WriteOnly Property FingerImage() As Bitmap
		Set(ByVal value As Bitmap)
			pbFingerprint.Image = value
		End Set
	End Property

	#End Region ' Public Properties

	#Region "Event Handling"

	Private Function CalculateDistance() As Single
		Return Convert.ToSingle(Math.Sqrt((_startPt.X - _endPt.X) * CDbl(_startPt.X - _endPt.X) + (_startPt.Y - _endPt.Y) * CDbl(_startPt.Y - _endPt.Y)))
	End Function

	Private Sub pbFingerprint_MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs) Handles pbFingerprint.MouseDown
		If e.Button = MouseButtons.Left Then
			_startPt = e.Location
			_endPt = Point.Empty
			pbFingerprint.Cursor = Cursors.Cross
			_dragging = True
		End If
	End Sub

	Private Sub pbFingerprint_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs) Handles pbFingerprint.MouseMove
		If e.Button = MouseButtons.Left AndAlso _dragging Then
			_endPt = e.Location
			pbFingerprint.Invalidate()
		End If
	End Sub

	Private Sub pbFingerprint_MouseUp(ByVal sender As Object, ByVal e As MouseEventArgs) Handles pbFingerprint.MouseUp
		If e.Button = MouseButtons.Left AndAlso _dragging Then
			_endPt = e.Location
			Cursor = Cursors.Arrow
			_dragging = False

			Dim scaleToInch As Single = 1.0f
			If rbCentimeters.Checked Then
				scaleToInch = 1.0f / 2.54f
			End If

			Dim distance As Single = CalculateDistance()
			If distance > 50.0 Then
				nudHorzResolution.Value = Convert.ToDecimal(distance / (Convert.ToSingle(nudUnitScale.Value) * scaleToInch))
				nudVertResolution.Value = nudHorzResolution.Value
			Else
				Utilities.ShowInformation("Please draw a longer line segment.")
			End If

			_startPt = Point.Empty
			_endPt = Point.Empty
			pbFingerprint.Invalidate()
		End If
	End Sub

	Private Sub DrawSelectionLine(ByVal g As Graphics, ByVal pen As Pen)
		' draw pointer markers
		g.DrawLine(pen, _startPt.X - EndMarkerSize, _startPt.Y, _startPt.X + EndMarkerSize, _startPt.Y)
		g.DrawLine(pen, _startPt.X, _startPt.Y - EndMarkerSize, _startPt.X, _startPt.Y + EndMarkerSize)
		g.DrawLine(pen, _endPt.X - EndMarkerSize, _endPt.Y, _endPt.X + EndMarkerSize, _endPt.Y)
		g.DrawLine(pen, _endPt.X, _endPt.Y - EndMarkerSize, _endPt.X, _endPt.Y + EndMarkerSize)
		' draw line between starting and ending points
		g.DrawLine(pen, _startPt, _endPt)
	End Sub

	Private Sub pbFingerprint_Paint(ByVal sender As Object, ByVal e As PaintEventArgs) Handles pbFingerprint.Paint
		If (Not _startPt.IsEmpty) AndAlso (Not _endPt.IsEmpty) Then
			Dim g As Graphics = e.Graphics

			Dim smoothingMode As System.Drawing.Drawing2D.SmoothingMode = g.SmoothingMode
			g.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality

			Using pen = New Pen(Color.White, 3)
				DrawSelectionLine(g, pen)
			End Using

			Using pen = New Pen(Color.Green, 1)
				DrawSelectionLine(g, pen)
			End Using

			g.SmoothingMode = smoothingMode
		End If
	End Sub

	Private Sub nudHorzResolution_ValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles nudHorzResolution.ValueChanged
		errorProvider1.SetError(nudHorzResolution,If(nudHorzResolution.Value < 250D, "Current resolution is lower than recommended minimum.", ""))
	End Sub

	Private Sub nudVertResolution_ValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles nudVertResolution.ValueChanged
		errorProvider1.SetError(nudVertResolution,If(nudVertResolution.Value < 250D, "Current resolution is lower than recommended minimum.", ""))
	End Sub

	#End Region ' Event Handling
End Class
